### ISSUE TITLE
[RFC] Proposal - Introduce public RFC process for C# features selection and prioritization for major C# releases

### ISSUE THREAD

**Issue by 4creators** [dotnet/csharplang#1838]
Following the example of open source communities working on compilers (F#, Rust) I would like to propose using open Request For Comments period for publicly communicating with C# language community about features which are going to be implemented by language, perhaps, about which features could be deprecated. 

The RFC should be started by language team making an open proposal for next major release and discussion should continue until the deadline set in RFC. Received feedback could be a very good basis besides other sources for decisions which are obviously reserved for LDM.

This form of public consultation is one of the best ways to reach consensus and build strong community support for compiler, language and language team. Everyone who may have chance to comment will exert real influence on the language development  just by following simple RFC process. Those who do not comment should not complain.

Related to #1600

**Comment by svick** [comment]
What exactly would be the difference between the proposed process and the current one, apart from changing the name from "championed proposal" to "RFC"?

**Comment by 4creators** [comment]
@HaloFour @svick 

Seems that I was not clear enough. It is not proposal to do RFC for single features it is proposal to put **list of features which are going to be included in next major release** (perhaps with some minor point releases). 

1. Features which are going to be included in RFC are described elsewhere.
2. Only small number of proposed features should and could ever land in the language - process of choosing which ones should be lucky winners should be based on RFC and obviously left for the final decision of LDM.

**Comment by HaloFour** [comment]
@4creators 

Versions are not defined by a list of features decided in advance.  They are comprised of the features that can be completed in the appropriate time frame.  To add a process which attempts to time box the team and force a given feature to be implemented in a "next version" I believe would be very counter to how the language is designed today and would be a detriment to the overall evolution of C#.  Many features *slated* for C# 8.0 have been in discussion and design since before C# 6.0.  And not a single one of those features is guaranteed to be released with C# 8.0 either.

**Comment by theunrepentantgeek** [comment]
Why change something that's _not already broken_?

If you want to make the argument for change, perhaps a good place to start would be to identify the problems you see with the current process.

**Comment by jnm2** [comment]
I thought the general consensus was that the current process was kind of broken, where decisions were made about C# 7 releases long before the community found out about them and way too late to change course?

**Comment by HaloFour** [comment]
@jnm2 

IMO that has been due to the periodic delays in releasing design meeting notes thus not giving the community a lot of (or any in some cases) time to review and comment.  The team seems to be much more regular and timely with those notes lately.

And I don't think it helps to have a process to try to corral proposals into a semblance of a release.  The design process shouldn't be time boxed like that.

**Comment by Joe4evr** [comment]
> it is proposal to put **list of features which are going to be included in next major release**

We already have this [right here on the repo](https://github.com/dotnet/csharplang/milestone/8).

**Comment by 4creators** [comment]
>  [right here on the repo](https://github.com/dotnet/csharplang/milestone/8)

Hmm, is it a joke or simple an attempt to avoid any serious discussion. Could you point me to the issue in which addition of features to the milestone is discussed?

**Comment by 4creators** [comment]
> The design process shouldn't be time boxed like that.

Sorry, but such argument says nothing. @HaloFour  Could you elaborate and be more specific.

**Comment by 4creators** [comment]
> the general consensus was that the current process was kind of broken

This consensus is really widespread and seems to be contested only by small group of most active contributors to the online discussions but not necessarily to language itself.

**Comment by 4creators** [comment]
> Why change something that's not already broken?

@theunrepentantgeek Why you think that process is not broken? Can you compare it with other languages  developed as foss? My experience and observations tell something very much different.

**Comment by HaloFour** [comment]
@4creators 

> Sorry, but such argument says nothing.

It says that the community doesn't (and shouldn't) tell the designers when features should be ready to be included in a release.

> This consensus is really widespread and seems to be contested only by small group of most active contributors to the online discussions but not necessarily to language itself.

First, any discussion on this repo represents a *tiny* minority of feedback that the language team receives.  There are, what, a couple of dozen people posting here?  And the people who have specifically posted that they want more control over the evolution of the language represents a minority of that minority.

The C# language is not a democracy.  It's been expressed pretty explicitly that the C# team remains the final arbiter as to what features will be added and when they will be added.

I get it, you want more control and influence.  The team doesn't want that.  It's their language.

> Why you think that process is not broken?

The onus of evidence lays on the individual making the claim.

**Comment by 4creators** [comment]
One more important issue: I do not think that anyone serious could say that C# development currently is based on benovalent dictatorship. Such argument was raised in #1600 thread but never really developed. 

Currently LDM runs the decision process and interacts with community via this repo. Why not improve this interaction?

One of the most precious and limited resource is developer time and using RFC to discuss it's allocation is something separate and different from discussing syntactic and semantic merits of proposed features.

Honestly, it's hard to understand that such situation is continuing for several years already and any attempt to discuss it is pushed away by small number of posting parties.

**Comment by 4creators** [comment]
> It says that the community doesn't (and shouldn't) tell the designers

Let LDM speak on the issue and I don't think that in situation in which you do not represent LDM officially and yet accept its dictatorship and the fact that your opinion is totally irrelevant and yet you say something. So why you are expressing it at all - its logically contradictory, right?

**Comment by 4creators** [comment]
> The C# language is not a democracy. 

And again - perhaps some more logical self constraint would help in such discussions.

**Comment by HaloFour** [comment]
@4creators 

> I do not think that anyone serious could say that C# development currently is based on benovalent dictatorship.

As opposed to what?  If you don't think it's "benevolent" you're welcome to that opinion.  But if that's based on your opinion that the team is spending their time and resources on priorities with which you don't agree, that doesn't invalidate that designation at all.  It's guaranteed that not everyone will agree with the direction that the team takes.  But it's their decision to make and they do so in the interest of the greater community.

> Let LDM speak on the issue and I don't think that in situation in which you do not represent LDM officially and yet accept its dictatorship and the fact that your opinion is totally irrelevant and yet you say something. So why you are expressing it at all - its logically contradictory, right?

They have, repeatedly.  This repo and the "champion" process is the result of years of trying to evolve the language in the open with community feedback.

The fact that they are the final arbiters doesn't mean that they aren't open to community feedback.  It's been demonstrated many times that they are willing to do so, both by accepting and implementing community proposals and pivoting design decisions based on comments.  It's also been demonstrated many times that they are willing to make decisions against what may be a majority of community feedback, at least from this repo.

The fact that we can have this conversation on this repo is an example of the team accepting community feedback.  They had already decided to move the discussion entirely to mailing lists.  It was the resulting discussion on the Roslyn forum that led them to change their mind.

**Comment by tannergooding** [comment]
Small syntactic sugar features get in because:
1. They are small
2. They are easy/quick to implement
3. They generally have a broad impact on maintainability, ease of reading/writing, etc
4. They often don't require a lot of discussion on the right syntax and how it fits into the rest of the language
5. They generally don't require coordination with the Runtime or Framework
6. etc

We, simultaneously, get one or two fairly big features per major release, with improvements on those big features in minor releases

The bigger the language feature (and all the ones you called out are huge) the longer it takes to design, prototype, implement, and test (with a full feedback cycle included). Even if we had a solid syntax for those already, with the full end-to-end design for Runtime, Framework, and Language support, you still likely wouldn't see them implemented for the next release.

**Comment by jaredpar** [comment]
@4creators 

> LDM stand up and discuss with us openly about roadmap and feature prioritization

The C# / VB compiler team has an status page that shows all of our work in progress, specs and optimistic release dates for all items:

https://github.com/dotnet/roslyn/blob/master/docs/Language%20Feature%20Status.md

As for the other part: "stand up and discuss with us openly". Consider some of the items we do today:

1. Provide speclets for all of our features: both on csharplang (for the language design) and roslyn (for the compiler implementation). These are sent via PRs to the various repositories and customers are encouraged, and quite frequently do, provide feedback. 
1. Provide notes for our design meetings within a calendar week of their completion. This is admittedly newer. In the past we were a bit slack about this and had no commitment to when they would be posted. But we got feedback an that and now have a team member dedicated to this task.
1. Every feature under consideration for the language is explicitly labeled in this repository (look for [Proposal Champion](https://github.com/dotnet/csharplang/labels/Proposal%20champion)). This is true for both features proposed by the C# team and the community. 
1. The community and C# team alike use issues on this repository to discuss new and old ideas alike. 

Your comments make it sound like we have some sort of closed process here which is simply not the case.

**Comment by HaloFour** [comment]
@4creators 

> In this way we end up with selection of small syntactic sugar features polluting language syntax 

C# 8.0 is also slated to ship with some big features.  Some of those features have been in design for 3+ years.  This includes non-nullable references, default interface methods and recursive pattern matching.  Those are all the result of conversations that started on CodePlex, long before they were mature enough to be considered "proposals".

> without any significant progress on the most pressing and important general features i.e. Type Classes, High Level Generics etc. etc.

What makes you think that the team isn't exploring those ideas?  There are several proposals on this repo, some opened by team members themselves, to discuss these concepts.  This will take time, a long time.  The team isn't going to halt all other design and implementation to chase these down their respective rabbit holes.

**Comment by MgSam** [comment]
I mostly agree with @4creators here. The language design process has minimal to zero community input on feature prioritization and it has always been his way. Prioritization is very different from offering feedback on the actual implementation/shape of the features themselves (where the community has a some input). The LDM decides prioritization in closed-door meetings and the feature status page are the breadcrumbs we get months after the fact telling us what will be in the next version.

If I'm on the Bing team, the Xamarim team, or the ASP.NET team, the LDM makes my features a priority and the community finds out about it months later. Let's not all forget the `in` keyword fiasco already. The LDM added **and shipped** a new keyword the language, which people find almost universally confusing, and told the community about it after the fact. You can't get a more textbook example of poor open source development than that.

Up until a few months ago they would even go for months-long periods without even posting LDM notes. It looks like Mads maybe, finally, has been relieved of this responsibility so that it gets done in a more timely fashion, but that certainly shouldn't suggest that the process is now perfect.

I'm surprised to see people who have been in this community for years so oblivious to the problems with the C# language design process. I guess when you've been ignored for long enough you start thanking people when they give you even the slightest bit of attention. 

@HaloFour It is disingenuous to say that they work on lots of features in parallel and they are just finished whenever they're finished. We both know that's not true. Like any team, there is a finite amount of work the design team can do- they choose where they are going to focus their work. This act of prioritization is and always has been completely opaque to outside input. To imply that because they post new feature suggestions that they don't have priorities is absurd. Does having a backlog of work items at your job mean that there are no priorities?

**Comment by 4creators** [comment]
> Your comments make it sound like we have some sort of closed process here which is simply not the case.

@jaredpar 

In contrary. Due to **partially open process** we have a small say in language design. **We do appreciate it very much**. However (pls look at @MgSam  above comment), what is missing, this is part of the discussion with community not about design details but about what to actually implement and when implement it. This is equally important to how we design feature.

[Status page](https://github.com/dotnet/roslyn/blob/master/docs/Language%20Feature%20Status.md) does not replace discussion which proceeds status page creation. What I am asking is to allow community to have such discussion in the form of RFC to avoid community member alienation and frustration. Let people or (developers) be heard (again basic democratic slogan 😄 ) even if that means only thought material for LDM.

IMHO nothing prevents language team to expand community participation in that simple but very important way.

**Comment by gafter** [comment]
We accept community contributions.  Open-source collaborators who want to see major features prioritized into earlier (rather than later) releases are welcome to do the work of designing the feature, convincing everyone that the design is right, and implementing that design.  When we assign releases to features we are just signaling how we are planning to spend our design+implementation resources.

**Comment by 4creators** [comment]
@gafter 

Thanks for fast reaction. 

> When we assign releases to features we are just signaling how we are planning to spend our design+implementation resources.

We appreciate that very much. However, again and again, **let us express our opinions on how community may want to prioritize** _desing+implementation resources_ including community contributions. This may help very much in decisions taken by individual developers choosing on which features and when work could be started and it may create much better coordination of community resources. IMHO community may contribute a lot.

How you will later decide to use community feedback is up to language team.

**Comment by CyrusNajmabadi** [comment]
As specified in the past:

> We will keep growing C# to meet the evolving needs of developers and remain a state of the art programming language. We will innovate aggressively, while being very careful to stay within the spirit of the language. Given the diversity of the developer base, we will prefer language and performance improvements that benefit all or most developers, avoiding over-focusing on a given segment. **We will continue to empower the broader ecosystem and grow its role in C#’s future, while maintaining strong stewardship of design decisions to ensure continued coherence.**

The last part is particularly relevant.

**Comment by CyrusNajmabadi** [comment]
> We appreciate that very much. However, again and again, let us express our opinions on how community may want to prioritize desing+implementation resources including community contributions. 

AFAICT, nothing is stopping anyone from expressing their opinions on anything here :)   The only time anything has been stopped is when behavior has reached hte level of "excessive trolling".  Otherwise, people are totally free to give any and all feedback on all things.  That includes:

1. priorities of things.  
2. design and impls of things. 
3. community contributions.

All these things are stuff that can be freely provided.  The team and LDM do look at this stuff.  What is not guaranteed though is that the team or hte LDM will agree with any of that feedback or do anything about it.  As above, they maintain stewardship here and all these are simply signals that feed into their decision making process which is *ultimately* entirely up to them.

**Comment by CyrusNajmabadi** [comment]
> In this way we end up with selection of small syntactic sugar features polluting language syntax without any significant progress on the most pressing and important general features i.e. Type Classes, High Level Generics etc. etc.

What you consider "the most pressing and important general features" is not a universal belief.  The goal of the LDM is to balance their view over *all the needs* they see out the for the community as a large.  And, what we've found in the past, is that a balance between small nice-to-have features, and large major-impact features, leads to decent language upgrades that benefit the ecosystem in a healthy way.  

Only doing small-syntactic-sugar features would likely mean keeping C# in a local maxima zone.  Whereas, only doing large-scale features would mean increased frustration over the years as people continued having to deal with the onerous syntax day in and day out.

By working on both, we can provide features that can significantly improve development in both spaces.

**Comment by 4creators** [comment]
> We will continue to empower the broader ecosystem and grow its role in C#’s future, while maintaining strong stewardship of design decisions to ensure continued coherence

**Yes, and nobody questions this**. Using a RFC process for gathering community feedback on prioritization does not undermine decision making freedom of  **The team and LDM** in smallest part.

> What is not guaranteed though is that the team or hte LDM will agree with any of that feedback or do anything about it.

And this is fine with me as I have stressed many times.

How about RFC in the following form:

### Request For Comments for C# major release v9.0 planned for 1H 2021 (this is example 😄 only)

#### The Team and LDM would like to receive community feedback on planned features implementation and deprecation for next major C# language release. Please provide feedback on our choices, your own proposals and 

1. Implement final design resulting from discussions on [Exploration: Roles, extension interfaces and static interface members](https://github.com/dotnet/csharplang/issues/1711) and [Exploration: Shapes and Extensions](https://github.com/dotnet/csharplang/issues/164)

2. Implement [Higher Kinded Polymorphism / Generics on Generics](https://github.com/dotnet/csharplang/issues/339)

3. Implement [Proposal: Overridable `default` operator](https://github.com/dotnet/csharplang/issues/167)

4. Structures as pseudo native classes with inheritance and explicit lifetime management.

5. Allow overriding of all C# operators and `new` keyword, add `delete` keyword.

6. Deprecate `unsafe` keyword and unsafe semantics.

#### THIS IS EXAMPLE NOT PROPOSAL

**Comment by 4creators** [comment]
> We are letting you express that. See, for example, your #1600. The closest thing to a consensus (by vote) list of proposed features I can see in that thread is #1600 (comment)

@gafter 

My proposal is not about allowing me to raise issue what I am doing, the RFC proposal is about The Team and LDM participation in the process without incurring any obligations on them. The example of the very requested RFC process which already took place is the following issue opened by @MadsTorgersen: [Triage of features in the C# 7.2 milestone. ... Let's discuss](https://github.com/dotnet/csharplang/issues/789)

Let's do this systematically in the form of RFC (or any other) what would be desired solution.

**Comment by gafter** [comment]
We don't need to assign RFC numbers, as github already assigns unique numbers to issues.

We don't need to have discrete times where we are open for feedback and then closed for feedback, as we are always open for feedback.  Comments are welcome at any time.  The Team and LDM would like to receive community feedback on planned features implementation at all times.

We won't suggest dates for language releases because lawyers.

I'm not sure I understand how your suggestion of adding a new RFC numbering scheme and being closed for feedback after a given date would improve things.

**Comment by theunrepentantgeek** [comment]
@4creators
> Why you think that process is not broken? 

Because it's *working* - we're seeing a regular cadence of language releases come out.

Is there room for improvement? ***Absolutely*** - there's room for improvement to **any process**. We're seeing improvements happening all the time. 

> Using a RFC process for gathering community feedback on prioritization does not undermine decision making freedom of The team and LDM in smallest part.

The term ***RFC*** is a very loaded term, with a lot of history. There are a lot of folks who see them as *binding* (**1**) and who therefore would react poorly when *perceived promises* were broken. This would happen regardless of your opinions of the process.

(**1**) Twenty years ago, I was one of those folks - until someone gently educated me otherwise.

**Comment by CyrusNajmabadi** [comment]
> Yes, and nobody questions this. Using a RFC process for gathering community feedback on prioritization does not undermine decision making freedom of The team and LDM in smallest part.

Could you explain what you mean exactly by "RFC process"?  Please, be specific.

**Comment by Joe4evr** [comment]
> Could you point me to the issue in which addition of features to the milestone is discussed?

There is no such discussion, because assigning a feature to a milestone is **entirely at the discretion of the LDM**. Not to mention that features can move forwards and backwards depending on what the team views is feasible to attain for the given release window.

**Comment by CyrusNajmabadi** [comment]
> How about RFC in the following form:
Request For Comments for C# major release v9.0 planned for 1H 2021 (this is example 😄 only)
The Team and LDM would like to receive community feedback on planned features implementation and deprecation for next major C# language release. Please provide feedback on our choices,

It's unclear to me why this would be an improvement.  Instead of having each of those issues tracked individually, there would now be multiple locations in which it would be happening.  

Furthermore, this would be *especially* bad on a forum like github.  That's because there is no proper threading.  So now instead of having 10 individual issues, each focused on one specific langauge change, you have one issue with people talking about 10 different thins.  That seems *really* bad.  It feels like it would be extremely difficult to keep things understandable.

Heck, when we have single issues *today* about a single language feature, it can often descend into complex conversations about subpieces of the single feature.  Even this is difficult to deal with.  So it seems *really* bad if this was compounded by having an RFC that merges these all into one thing.

**Comment by MadsTorgersen** [comment]
This is a great discussion, and it's been really enlightening and encouraging to see the different viewpoints. Thanks all!

As you know we've been fiddling with this process over the past several years, and are still making improvements. One recent improvement was getting a dedicated notetaker for the LDMs, who can communicate the outcomes back to the community with much more detail, accuracy and expediency than I was able to.

Language innovation is a very fluid process, and very hard to box in. For every single proposed feature there is an unpredictable amount of design work, challenges and interactions with other features, as well as a high risk of failure, in that we often cannot find a design that lives up to our standards. The aspect of _how much we want to add the feature_ can and certainly will factor into how much time we allocate to the design process for it, but assigning it to a release even tentatively comes with great uncertainty.

Through trial and error we've landed on the milestone system for championed features. This process commits us to looking at a championed feature within a given timeframe. Oftentimes, though, "looking at it" leads to discovering that it's either less desirable than competing features or harder than we thought. Then we'll either remove it or kick the can down the road to a later release. This is far from an exact science, but it's actually working pretty well. _Every artifact of this process is publicly visible on this repo_. We don't do the planning "somewhere else" and post it here when we care to. This is our working space, right here. If you want to see what is being considered for C# 8.0, just look at the [milestone](https://github.com/dotnet/csharplang/issues?q=is%3Aopen+is%3Aissue+milestone%3A%228.0+candidate%22+label%3A%22Proposal+champion%22).

It is the responsibility of the LDM to ensure that we not only build the features right, but also build the right features. Obviously, "right" is in the eye of the beholder. We try to take an integrative view across the current needs of the whole community as well as the long term success of C# and the .NET platform, as we can best discern it. We all want C# to solve our problems today, and we also want it to be desirable and successful into the future: we're all to some extent betting _our_ future on it. Last year I tried to formulate the philosophy behind our decisions in the [.NET Language Strategy](https://blogs.msdn.microsoft.com/dotnet/2017/02/01/the-net-language-strategy/), so that the decision making framework _itself_ is also out in the open.

So now you can see _what_ decisions we make, as well as _how_ we make them, and _why_ we make them, _when_ we make them. When you don't like them you can influence us right here.

**Comment by 4creators** [comment]
@MadsTorgersen @gafter @jaredpar @CyrusNajmabadi 

Thank you all for thorough discussion. IMHO we got to the core of the problem and it seems that the best way to reconcile different needs and expectation related to: 

(i) flexibility and unpredictability of development and thread/feature based issue tracking system and nature of design process on one side and,
(ii) need of the inclusive role for the community on the other side 

**could be small change to current practice, actually, not a change but continuation of already existing practice**.

The triage of milestones done at the LDM meeting and here https://github.com/dotnet/csharplang/issues/789 discussed with the community is a great example. It will be very helpful if such process would continue.

Triage of the features with request for feedback as was done by

**Comment by MadsTorgersen** [comment]
@4creators Thanks! Yes, the triage happens in C# LDM, and so the design notes should generally bear that out. We're due for a new round of triage, but for a couple of weeks I've had to push it out from the agenda due to issues that were blocking developers.

I think we'll get to it within a couple of weeks, and then the result will be in the design notes.

Now that we're on the topic, the agendas are generally very late bound, especially at times like now where a lot of features are mostly figured out (for C# 8.0) and developers run into unresolved issues that we need to deal with quickly to keep them productive. When I know the agendas in advance, I sometimes put them up in the form of ahead-of-time design notes containing a proposed agenda.

**Comment by HaloFour** [comment]
@MadsTorgersen 

Speaking of triage, I am concerned with the growing number of proposals that seem to be stuck in limbo.  They're not "good enough" to be championed but also not bad enough to warrant being closed outright.  It's be nice if those issues could see a little love from the C# team.

I am sure that free time is something in short supply, but I think it would be useful if perhaps some time per month could be dedicated to a couple of these proposals during C# LDM.  Perhaps they could be selected in order, or team members could "adopt" ones that they feel lukewarm about.  Then the team could gauge a relatively quick level of effort and degree of interest.  Those proposals that don't seem like they would ever really be considered could be closed and those that remain could perhaps be put up for community consideration as "up for grabs" or perhaps to a vote that could set a priority/order in which they might be implemented in minor releases.

Lastly, while there is certainly interest in the community to actually try to implement a proposal, and a few success stories, I think we can all recognize that the barrier to entry is insanely high.  Again, with free time in short supply I recognize that this might be difficult to implement but I could see value in a "community ambassador" program where members of the C# team help community members through the process of actually trying to make the necessary changes to Roslyn in order to implement new features.

Consider this all 🍝 against the wall.

**Comment by theunrepentantgeek** [comment]
Very good point @HaloFour - I made a couple of proposals that didn't go anywhere, and I just left them open. I'll go back and reconsider whether they should be closed or not.

**Comment by Joe4evr** [comment]
> I'd also add that anything that requires CLR or CoreFX changes is outright impossible for the community to wholly implement in our own time

Tell that to the guy that's worked on implementing generic attributes with [an associated CLR patch](https://github.com/dotnet/coreclr/pull/9189). :upside_down_face:

**Comment by yaakov-h** [comment]
@Joe4evr and the desktop CLR?

**Comment by DavidArno** [comment]
@yaakov-h,

This may be a controversial view, but I think it time to retire the framework CLR. The future is with dotnetcore. What really is the benefit with MS maintaining two separate CLRs?

This thread is an interesting one for me. A year ago, I'd have up-voted @4creators comments and rallied against the current process. But I've come to appreciate the way it works recently. This forum is dominated by a handful of opinionated folk. There's nothing wrong with being opinionated like this, but we represent a tiny fraction of the C# user-base. If this forum were the course by which language features were chosen, then we'd end up with a language highly skewed toward what that small vocal group wanted, at the expense of everyone else.

A case in point is scope leakage with "out vars". Many of us (possibly the majority) here were strongly opposed to this and ranted at the LDM for ignoring us. Some - especially myself - predicted mass confusion when the feature was released due to the different scoping rules. But the feature has been very well received by the user base as a whole. Folk I've spoken to aren't really interested in strict language rules etc. What they want is features that "feel right". Out vars "leak" in just the right way for them: it makes sense and makes the feature usable. The LDM made the right decision. They listened to us, but ultimately disagreed and took a different path that turned out to be the right one. If we had a process as described in the OP though, this wouldn't have happened and many users would have been unhappy with the results.

I still think things could be improved here. Dealing with the hundreds of proposals that are stuck in limbo, having neither been championed on rejected, being the main one. But I feel the process around how features adopted by the LDM are discussed, prototyped and eventually released does work well. I don't think it is broken and doesn't need radically changing.

**Comment by jnm2** [comment]
@yaakov-h I understood that the desktop CLR was actually built from the same sources or merged the same sources, but I could be wrong.

**Comment by chucker** [comment]
>It's been expressed pretty explicitly that the C# team remains the final arbiter as to what features will be added and when they will be added.

As much as I respect that (I don't believe a fully democratic process leads to a good design), I noticed something a few posts further down:

>https://github.com/dotnet/roslyn/blob/master/docs/Language%20Feature%20Status.md

I find this document [at its current version](
https://github.com/dotnet/roslyn/commit/92e197400a9007901f5f33bb8455461aa45be376) a bit puzzling because it appears to tell a slightly different story than the recent [Building C# 8.0](https://blogs.msdn.microsoft.com/dotnet/2018/11/12/building-c-8-0/) blog post. That blog post names the following "most significant features slated for C# 8.0":

Nullable reference types
Async streams
Ranges and indices
Default implementations of interface members
Recursive patterns
Switch expressions
Target-typed new-expressions

It stands to reason, then that the **State** column in the Language Feature Status document should reflect that these seven features have a common, fairly advanced state. However:

* only two of those are in "Merged to dev16 preview1" state
* three are shown as "Prototype" (shouldn't they at least be *In Progress*?)
* one (Switch expressions) doesn't appear to be mentioned at all
* the status document lists two additional "Merged to dev16 preview1" features that aren't mentioned in the blog post (perhaps because they are considered minor?)

In turn, the 8.0 candidate milestone gives yet another list. (That list also doesn't appear to include Switch expressions. Maybe my google-fu is being poor here.) The name of the milestone also suggests that essentially _no_ feature is currently a lock for 8.0, which surely isn't true, given that some have already been merged into VS16? Why not move those into a separate 8.0 (non-candidate) milestone?

Lastly, the blog post also links videos that hype features that probably(?) won't make it into 8.0, such as extension everything.

That's four different somewhat overlapping, somewhat conflicting sources on what will or won't make it into 8.0. I _assume_ the blog post is the most canonical source for the time being.

This may sound like nitpicking, but ultimately, I don't find it very transparent. It would help a lot if those documents were more in sync, so the community would know what the team is planning (vs. what is merely a wishful best-case scenario), and could give feedback accordingly.

**Comment by gafter** [comment]
Until we announce what future version of VS corresponds to what future version of C#, please do not assume. Also, please do not assume release dates or how advanced the implementation "should be" compared to your imagined dates.

No feature is a lock for C# 8.0 until it has shipped.

The document https://github.com/dotnet/roslyn/blob/master/docs/Language%20Feature%20Status.md describes what the compiler team is working on, not what language features are expected. However, there is obviously a lot of overlap. For the purposes of both specification and implementation, switch expressions are being treated as part of recursive pattern-matching, which is why it doesn't get a separate entry. But it is highlighted separately in the blog post because it will feel significant.

You should not be surprised that previous, earlier estimates about the contents of C# 8.0 are different from more recent ones.
